pattern/src/lib.rs

```diff
@@ -1,9 +1,107 @@
+//! - patterns which correspond to bitwise cells, such as [`PAnd`], can match
+//!   a [`Value`] even if it is made up from [`Net`]s coming from many distinct
+//!   [`And`] cells.
+//! - specialized patterns can be easily implemented, such as [`PPow2@a`],
+//!   which matches a constant power of 2 and captures the exponent into `a`.
+//!
+//! [`And`]: prjunnamed_netlist::Cell::And
+//! [`Net`]: prjunnamed_netlist::Net
+//! [`Value`]: prjunnamed_netlist::Value
+//! [`PAnd`]: patterns::PAnd
+//! [`PPow2@a`]: patterns::PPow2
+//! [`PZExt`]: patterns::PZExt
+//!
+//! By convention, the names of patterns are prefixed with `P`.
+//!
+//! The general netlist match syntax looks as follows:
+//!
+//! ```no_run
+//! # use prjunnamed_netlist::{Design, Net, Value};
+//! # use prjunnamed_pattern::netlist_matches;
+//! # use prjunnamed_pattern::patterns::*;
+//! let matcher = netlist_matches! {
+//!     [PAdc@y [PAny@a] [PZero] [PInput@c ("cin")]] => (y, a, c);
+//! };
+//! # let design: Design = todo!();
+//! # let target: Value = todo!();
+//! # let result: Option<(Value, Value, Net)> = matcher(&design, &target);
+//! ```
+//!
+//! A pattern call is delimited with square brackets and consists of the
+//! following parts:
+//! - the name of the pattern being called, e.g. `PAdc`
+//! - a result capture, e.g. `@y`. This is optional, and if it is not provided,
+//!   the value captured by the pattern is discarded.
+//!
+//!   Each pattern captures exactly one value. Often, it is the [`Net`] or
+//!   [`Value`] it is being matched against. In rare cases, the captured value
+//!   is of the unit type `()`, in which case it is usually discarded by the
+//!   user.
+//! - a number of expression arguments, e.g. `("cin")`. Expression arguments
+//!   are evaluated according to normal Rust rules, and the results are passed
+//!   to the pattern. Each pattern has a fixed number of expression arguments
+//!   it expects.
+//!
+//!   If multiple expression arguments are being passed to the pattern, each
+//!   requires its own pair of parentheses, e.g. `PFeline@cat ("meow") (42)`
+//! - a number of pattern arguments, e.g. `[PZero]`. Pattern arguments
+//!   follow the pattern call syntax being described.
+//!
+//!   Each pattern has a fixed number of pattern arguments it expects.
+//!   Most commonly, they correspond to the inputs of the cell being matched.
+//!
+//!   If you do not wish to perform further matching on the inputs, use `[PAny]`.
+//!
+//! Much like `match`, [`netlist_matches!`] supports `if` and `if let` guards.
+//! For example:
+//!
+//! ```no_run
+//! # use prjunnamed_netlist::{Design, Net, Value};
+//! # use prjunnamed_pattern::netlist_matches;
+//! # use prjunnamed_pattern::patterns::*;
+//! let matcher = netlist_matches! {
+//!     [PXor [PAny@a] [PAny@b]] if a == b => Value::zero(a.len());
+//! };
+//! # let design: Design = todo!();
+//! # let target: Value = todo!();
+//! # let result: Option<Value> = matcher(&design, &target);
+//! ```
```

Comment:
You refer to an "applicable rule" below but I don't think it's ever properly introduced. We should introduce concepts before referring to them (and I suppose we should also decide this: while the things you put between `{}` in `netlist_replace!` are clearly "rules", does this apply to `netlist_matches!`?